File: dad-is-great-in-C/ognfunc.cpp
// a list of useful functions written by me

#include "ognfunc.hpp"
#include <iterator>
#include <string>

unsigned long long numsum(unsigned long long n) { // sum of number digits
  auto res = 0ULL;
  while (n) {
    res += n % 10;
    n /= 10;
  }
  return res;
}

unsigned long long reverse(unsigned long long n) { // reverse number
  auto rev = 0ULL;
  while (n) {
    rev = rev * 10 + (n % 10);
    n /= 10;
  }
  return rev;
}

unsigned long long numsys(unsigned long long n, unsigned char k) {
  auto ns = 1ULL;
  while (n) {
    ns = ns * 10 + (n % k);
    n /= k;
  }
  return (reverse(ns) - 1) / 10;
}

unsigned short numcount(unsigned long long n) { // counts number digits
  unsigned short count = 0;
  while (n) {
    n /= 10;
    count++;
  }
  return count;
}

std::string numsys16(unsigned long long n, unsigned char k) {
  const char NUMS[16] = {'0', '1', '2', '3', '4', '5', '6', '7',
                         '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
  std::string res = "";
  while (n) {
    res += NUMS[n % k];
    n /= k;
  }
  std::string rev(res.rbegin(), res.rend());
  return rev;
}

long double binpow(long double b, unsigned short e) { // binary power
  auto v = 1.0L;
  while (e != 0) {
    if ((e & 1) != 0)
      v *= b;
    b *= b;
    e >>= 1;
  }
  return v;
}

File: pie/ognfunc.py
# a list of useful functions, like it did in dad-is-great-in-C


def rev(n):
    rev = 0
    while n:
        rev = rev * 10 + (n % 10)
        n //= 10
    return rev


def numsys(n, k):
    ns = ""
    while n:
        ns += str(n % k)
        n //= k
    return int(ns[::-1])


def numsys16(n, k):
    nums = {
        0: "0",
        1: "1",
        2: "2",
        3: "3",
        4: "4",
        5: "5",
        6: "6",
        7: "7",
        8: "8",
        9: "9",
        10: "A",
        11: "B",
        12: "C",
        13: "D",
        14: "E",
        15: "F",
    }
    res = ""
    while n:
        res += nums[n % k]
        n //= k
    return res[::-1]
